### PR TITLE
feat: I18n

### PR DESCRIPTION
**Pull request by tommoor**
Integration branch for https://github.com/outline/outline/pull/1624
closes https://github.com/outline/outline/issues/880

image

**Additional translation and i18n work should be PR'd into this branch**

### CODE REVIEW

**Comment by tommoor** [comment]
@glaand the frustrating thing right now is that flow still doesn't understand decorators, which means we have to use the hoc either in the non-decorator syntax or manually type `t` in every component that uses it, the former is probably the least of two evils

**Comment by tommoor** [comment]
@glaand I also have a question about the tests – it seems like they are enforcing that all strings in the default language exist in all languages? This seems wrong unless the default language gets copied across to all languages when missing, we don't want to be in a situation where you can't add new text unless it's translated into all languages.

**Comment by glaand** [comment]
> 
> 
> @glaand I also have a question about the tests – it seems like they are enforcing that all strings in the default language exist in all languages? This seems wrong unless the default language gets copied across to all languages when missing, we don't want to be in a situation where you can't add new text unless it's translated into all languages.

You are right. The reason i did this for now was to ensure that the languages were working fine without checking every string manually. 

Do you want to remove this tests or should i do it? @tommoor

File: shared/translations/i18n.js [hunk 1]
```diff
@@ -20,18 +32,10 @@ const initI18n = () => {
...
-      },
-    },
+    resources,
   });
 };
 
-export { initI18n, i18n, en_US, de_DE, pt_PT };
+const languages: string[] = Object.keys(resources);
+
+export { initI18n, languages, i18n, en_US, de_DE, pt_PT };
```
Comment (glaand) [inline shared/translations/i18n.js:41]: Very good one here :P @tommoor

File: server/models/User.js [hunk 2]
```diff
@@ -39,6 +40,9 @@ const User = sequelize.define(
     language: {
       type: DataTypes.STRING,
       defaultValue: process.env.DEFAULT_LANGUAGE,
+      validate: {
+        isIn: [languages],
```
Comment (glaand) [inline server/models/User.js:44]: So i guess this does the auto check of user.language is in languages @tommoor ?

Comment (tommoor) [reply server/models/User.js:44]: Yep, validation at the ORM level a bit more resilient. It throws a specific error type which returns the right error to the client.

**Comment by tommoor** [comment]
**Do you want to remove this tests or should i do it? @tommoor**

Tests removed 👍

**Comment by tommoor** [comment]
I think the main thing I'd like to figure out before merging this first pass is what the contribution flow looks like – what is a good way to work through missing translations? 🤔

**Comment by glaand** [comment]
> 
> 
> I think the main thing I'd like to figure out before merging this first pass is what the contribution flow looks like – what is a good way to work through missing translations? 🤔

Right, better look it now. I've made a prototype of the future translation process :P , please take a look.

What do you think?

Please follow from process start til process end

Translation Process - Outline

**Comment by tommoor** [comment]
Thanks for the diagram, that's very useful. The "push" step can certainly happen from the `ci` servers, perhaps in the future the "pull" step could be an automated PR too.

It's a shame that localize doesn'ts seem to offer a free plan for open source projects for their TMS seeing as it seems to be well integrated with react-i18n

**Comment by glaand** [comment]
> 
> 
> Thanks for the diagram, that's very useful. The "push" step can certainly happen from the `ci` servers, perhaps in the future the "pull" step could be an automated PR too.
> 
> It's a shame that [locize](https://locize.com/) doesn't seem to offer a free plan for open source projects for their TMS seeing as it seems to be well integrated with react-i18next. Still, it's cheap so that might be the one to start with.

Yeah true. A bad business move but anyways... It looks like ABB uses locize, ABB its a well known company here in Switzerland (In the billions category) so i guess it's a good referal to take them.

Looks like it costs 5 dollars per month. How do you handle expenses at Outline @tommoor 

I guess we could choose locize for this yes

**Comment by tommoor** [comment]
$5 is base price and charges per-word above that, but as long as it integrates well it's fine. I will setup an account that the Outline SaaS business will sponsor.

**Comment by glaand** [comment]
> 
> 
> $5 is base price and charges per-word above that, but as long as it integrates well it's fine. I will setup an account that the Outline SaaS business will sponsor.

Alright cool thank you very much :)

**Comment by tommoor** [comment]
I split the translations out from the js bundle to reduce initial load time, they're now served via HTTP. I'm happy to merge this now but we do still need a strategy for identifying which translations are needed

**Comment by glaand** [comment]
@tommoor Thank you for the changes. Yes indeed we need!

Im gonna take a look at it today, then i will post a proposal here.

**Comment by MatheusRV** [comment]
Is this ready for new languages? I can put in PT BR, if yes.

**Comment by tommoor** [comment]
It is, however this branch will be merged to `develop` very soon. Translations should now be provided through the portal, see this file for instructions:

https://github.com/outline/outline/blob/8aa215dce350f209481444ee699e4e6da25fa214/CONTRIBUTING.md

Thanks!

**Comment by MatheusRV** [comment]
@tommoor I can't add new languages.

image

**Comment by tommoor** [comment]
Okay added pt_BR just now 😌

**Pull request merged**